### DISCUSSION TITLE
Respect tsconfig's "paths" option?

### DISCUSSION THREAD

**Discussion by theDanielJLewis** [shellscape/jsx-email#129] [Q&A]
I'm a react-email refugee and I'm having some trouble. (First, it'd be great if there was a react-email migration guide!)

I have the following in my project's tsconfig.json:

```js
  "compilerOptions": {
    "paths": {
      "@/*": ["./src/*"]
    },
  },
```

JSX email doesn't like having `@/` in my import statements. Even though I can replace those with `../`, the preview still gives me "Failed to load url" errors for every `import from '@/…'` in my imported components.

I dread the idea of going through my entire project to replace the `@/` imports with relative paths. Is there a way to work around this in JSX email?

**Comment by shellscape** [comment]
Is this strictly with the preview app, or does that happen with `render` as well?

**Reply by theDanielJLewis** [reply]
It definitely happens in the preview app, but I can't tell if it happens with `render` because that CLI command spits out a bunch of new errors that are probably react-email migration steps I have to work through.

**Reply by shellscape** [reply]
OK step 1 is to get `render` working. I'll be in discord for a bit if you need some live help. 

The paths thing in the preview is going to be tricky because we're already using `@` as an alias within the Vite config https://github.com/shellscape/jsx-email/blob/31702c9490db666d623b3020ffa1ee24dd993825/packages/jsx-email/src/cli/commands/preview.ts#L63

This isn't the first time the tsconfig paths thing has come up. The general guidance there is that email templates shouldn't be reusing client components because client components aren't written for email, and the requirements are vastly different. Component reuse in email templates is common, but not common in a complex file tree. And the two folks that have raised the issue have just switched to using parent paths (e.g. `..`).

Now, there's https://www.npmjs.com/package/vite-tsconfig-paths. But we've been trying to avoid that because of the complexities it introduces and the way we have vite loading the base preview app from node_modules, and then using your template directories as root. There's also the issue of how to know which tsconfig file in the repo is the one to use. (And it has [https://github.com/aleclarson/vite-tsconfig-paths/issues](issues)). We could guess, but that would certainly result in some edge cases. That would almost certainly mean that we'd have to use a config file or CLI flags. There's been a call or two for allowing users to set Vite options, but that's a big ol can of worms.

So it boils down to juice vs squeeze. The complexity of adding support for tsconfig paths far outweighs using `../`. All that said I'm open to a compelling argument for it if there's a quick way to implement support.

**Reply by theDanielJLewis** [reply]
From my very limited understanding of the problem, could a solution be to simply _not_ use `'@': targetPath` in your Vite config, and then allow developers to use it?

**Reply by shellscape** [reply]
I don't believe so. Vite (esbuild under the hood) won't know what to do with the custom paths unless you tell it.

**Reply by theDanielJLewis** [reply]
So for the component issue, I should be okay if I just use relative paths for all problematic components?

**Reply by shellscape** [reply]
Yes that will get you unblocked. I'd like to work with you on a more palatable solution for the future though.

**Comment by shellscape** [comment]
I have a working solution for this and will be posting a PR shortly. Will tag you in that.

**Comment by adaptive-shield-matrix** [comment]
I have tried using jsx-email.
Not being able to use absolute typescript imports is a deal breaker for me.

"The general guidance there is that email templates shouldn't be reusing client components because client components aren't written for email, and the requirements are vastly different." 
-> does not apply because I only want to import pure typescript functions and variables that have nothing to do with React components. I want to be able to access many project/business specific variables and settings: site settings, business rules, classNames, i18n, magic numbers, etc... 

Using them as is - is not possible because of transitive imports. 
I can change and use relative imports in all jsx-emails code files, but I can't change all imports from absolute to relative across the rest of the 100k codebase.

Blocking and preventing the usage of one of the most common absolute import alias "@" in any project wanting to use jsx-email does not seem like a good idea.

**Reply by shellscape** [reply]
Did you miss my reply directly above yours?
